File: 2022/Javascript/src/Day4.js
function main() {
	const path = require('path')
	const main = require(path.join(path.resolve(''), 'Main.js'))
	const os = require('os')
	const lines = main.getDayInput(4).split(os.EOL)

	var useless = 0
	var overlapping = 0
	for (i in lines) {
		if (lines[i].length < 2) {
			continue
		}

		const nums = lines[i].split(/[\-,]/)
		const start1 = Number(nums[0])
		const end1 = Number(nums[1])
		const start2 = Number(nums[2])
		const end2 = Number(nums[3])
		if ((start1 >= start2 && end1 <= end2) || (start1 <= start2 && end1 >= end2)) {
			useless++
		}

		if ((start1 >= start2 && start1 <= end2) || (start2 >= start1 && start2 <= end1)) {
			overlapping++
		}
	}

	console.log(`There are ${useless} completely overlapping assignments.`)
	console.log(`There are ${overlapping} partially overlapping assignments.`)
}

main()

File: 2022/Cpp/src/Day4.cpp
/*
 * Day4.cpp
 *
 *  Created on: Dec 4, 2022
 *      Author: ToMe25
 */

#include "Main.h"

std::pair<std::string, std::string> day4Combined(std::ifstream input) {
	uint16_t useless = 0;
	uint16_t overlapping = 0;

	std::string line;
	while (std::getline(input, line)) {
		if (line.length() == 0) {
			continue;
		}

		const size_t first = line.find('-');
		const size_t comma = line.find(',', first);
		const size_t second = line.find('-', comma);

		const int start1 = std::stoi(line.substr(0, first));
		const int end1 = std::stoi(line.substr(first + 1, comma - first - 1));
		const int start2 = std::stoi(
				line.substr(comma + 1, second - comma - 1));
		const int end2 = std::stoi(line.substr(second + 1));
		if (start1 >= start2 && end1 <= end2) {
			useless++;
		} else if (start1 <= start2 && end1 >= end2) {
			useless++;
		}

		if ((start1 >= start2 && start1 <= end2)
				|| (start2 >= start1 && start2 <= end1)) {
			overlapping++;
		}
	}

	return {std::to_string(useless), std::to_string(overlapping)};
}

bool d4c = aoc::registerCombined(4, &day4Combined);
